Output locals mapping in trace logging

OddBloke opened the pull request:
The Locals line in the Frame Object output is now, for example:

Locals: {"num": RefCell { value: [PyObj int 9] }, "f": RefCell { value: [PyObj function] }}

(Note that in cases where the inner values of the PyObjectRefs are unexpected, I simply output to the log and continue; potentially we should panic instead, but I wasn't sure if that was appropriate in a trace log.)

windelbouwman commented on vm/src/frame.rs:
```diff
@@ -114,11 +114,16 @@ impl fmt::Debug for Frame {
-            .join(""); */
+        let local_str = match self.locals.borrow().kind {
+            PyObjectKind::Scope { ref scope } => match scope.locals.borrow().kind {
+                PyObjectKind::Dict { ref elements } => format!(" {:?}", elements),
```
Please use here the `.map` magic like in the value stack to print all locals below eachother.

windelbouwman commented on vm/src/frame.rs:
```diff
@@ -114,11 +114,16 @@ impl fmt::Debug for Frame {
+            PyObjectKind::Scope { ref scope } => match scope.locals.borrow().kind {
+                PyObjectKind::Dict { ref elements } => format!(" {:?}", elements),
+                ref unexpected => format!(
+                    "locals unexpectedly not wrapping a dict! instead: {:?}",
```
A panic is fine at this point I would say, since this is internally managed. If this is wrong, we are in really bad shape. It is not a python exception, but really a rust panic. Also consider `if let` for `match` statements with one arm only. It might be a good point to use here.

OddBloke commented:
Thanks for the comments. I'm away from my laptop today, so I'll address
them tomorrow.

On Wed, Aug 1, 2018, 11:53 Windel Bouwman, <notifications@github.com> wrote:

> *@windelbouwman* commented on this pull request.
> ------------------------------
>
> In vm/src/frame.rs
> <https://github.com/RustPython/RustPython/pull/22#discussion_r206934524>:
>
> > @@ -114,11 +114,16 @@ impl fmt::Debug for Frame {
>              .map(|elem| format!("\n  > {:?}", elem))
>              .collect::<Vec<_>>()
>              .join("");
> -        let local_str = "".to_string(); /* self.locals
> -            .iter()
> -            .map(|elem| format!("\n  {} = {}", elem.0, elem.1.borrow_mut().str()))
> -            .collect::<Vec<_>>()
> -            .join(""); */
> +        let local_str = match self.locals.borrow().kind {
> +            PyObjectKind::Scope { ref scope } => match scope.locals.borrow().kind {
> +                PyObjectKind::Dict { ref elements } => format!(" {:?}", elements),
> +                ref unexpected => format!(
> +                    "locals unexpectedly not wrapping a dict! instead: {:?}",
>
> A panic is fine at this point I would say, since this is internally
> managed. If this is wrong, we are in really bad shape. It is not a python
> exception, but really a rust panic. Also consider if let for match
> statements with one arm only. It might be a good point to use here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RustPython/RustPython/pull/22#pullrequestreview-142443947>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD1EOUzX9sHKtzhnmZ7pqOQk37VpnEuks5uMc7ngaJpZM4VpzQu>
> .
>

OddBloke commented:
Hi @windelbouwman, I've addressed your comments; the output format now looks like:

```
 Locals:
  g = <func>
  foo = <func>
  f = <func>
```

and we're panicking instead of simply logging. I didn't switch to `if let` because we capture the value of the default clause for use in the panic message, which is more readily expressed in a match

windelbouwman commented:
Hi @OddBloke , fantastic! Great job!
